### PR TITLE
perfetto_cmd: protect persistent traces against duplicate session names

### PR DESCRIPTION
**Pull request by safayat-google**
Use advisory file locks (flock) on persistent .tmp trace files to prevent duplicate
sessions with the same name from deleting or corrupting traces of an active session.

- Hold LOCK_EX on the persistent trace file while session is active.
- Skip unlinking and fail startup if an active session holds the lock.
- Skip reboot upload if file is locked by an active session.
- Add unit tests for duplicate session rejection and stale cleanup.

Bug: 382209797
Test: perfetto_unittests --gtest_filter="PerfettoCmdlineUnitTest.*"

### CODE REVIEW

File: src/perfetto_cmd/perfetto_cmd_android.cc [hunk 1]
```diff
@@ -302,6 +307,29 @@ void PerfettoCmd::WaitForPreviousRebootTraceUpload(
...
 
+  if (!base::FileExists(target_file_path)) {
+    return base::OkStatus();
+  }
+
+  // File exists after waiting for reboot task. Check if an active session is
+  // holding the lock before doing any cleanup.
+  {
+    base::ScopedFile existing_fd =
+        base::OpenFile(target_file_path, O_RDWR | O_CLOEXEC);
```
Comment (primiano) [inline src/perfetto_cmd/perfetto_cmd_android.cc]: base::OpenFile already takes care of CLOEXEC

File: src/perfetto_cmd/perfetto_cmd_android.cc [hunk 2]
```diff
@@ -302,6 +307,29 @@ void PerfettoCmd::WaitForPreviousRebootTraceUpload(
     }
   } while (cur_prop.empty() && base::GetBootTimeNs() < deadline);
 
+  if (!base::FileExists(target_file_path)) {
+    return base::OkStatus();
+  }
+
+  // File exists after waiting for reboot task. Check if an active session is
```
Comment (primiano) [inline src/perfetto_cmd/perfetto_cmd_android.cc]: I'd reword this comment as:

// This file might be lefover by a crashed instance of perfetto_cmd

File: src/perfetto_cmd/perfetto_cmd_android.cc [hunk 3]
```diff
@@ -302,6 +307,29 @@ void PerfettoCmd::WaitForPreviousRebootTraceUpload(
...
+      if (flock(existing_fd.get(), LOCK_EX | LOCK_NB) != 0) {
+        if (errno == EWOULDBLOCK || errno == EAGAIN) {
+          return base::ErrStatus(
+              "A persistent tracing session for '%s' is already running (%s is "
+              "locked).",
+              session_name.c_str(), target_file_path.c_str());
+        }
+        PERFETTO_PLOG("Failed to check flock on %s", target_file_path.c_str());
+      }
+      existing_fd.reset();
```
Comment (primiano) [inline src/perfetto_cmd/perfetto_cmd_android.cc]: This is not neeed given you are using scopedfile

File: src/perfetto_cmd/perfetto_cmd_android.cc [hunk 4]
```diff
@@ -284,15 +288,16 @@ base::ScopedFile PerfettoCmd::CreateUnlinkedTmpFile() {
   return fd;
 }
```
Comment (primiano) [inline src/perfetto_cmd/perfetto_cmd_android.cc:290]: at this point woudl be nice to have a comment explaining all the possible scenarios and what they return, as this function now is a bit more complex

**Review by primiano — APPROVED** [review]
LGTM with a few comments

File: src/perfetto_cmd/perfetto_cmd_android.cc (hunk 1 above)
Comment (safayat-google) [reply src/perfetto_cmd/perfetto_cmd_android.cc]: Removed the redundant `CLOEXEC`.

File: src/perfetto_cmd/perfetto_cmd_android.cc (hunk 2 above)
Comment (safayat-google) [reply src/perfetto_cmd/perfetto_cmd_android.cc]: Comments are updated.

The following block handles early return case when another instance of perfetto_cmd is still actively using it. Added your suggested comments where we push the `kRebootTraceUploadLeftover` atoms.

File: src/perfetto_cmd/perfetto_cmd_android.cc (hunk 3 above)
Comment (safayat-google) [reply src/perfetto_cmd/perfetto_cmd_android.cc]: Removed the explicit `reset` call.

File: src/perfetto_cmd/perfetto_cmd_android.cc (hunk 4 above)
Comment (safayat-google) [reply src/perfetto_cmd/perfetto_cmd_android.cc:290]: Thanks. Added comments explaining all scenarios.
